File: backend/app/services/audit_service.py
from app.models.audit_log import AuditLog
from app.models.user import User
from app.repositories.audit_repository import AuditLogRepository
from app.schemas.user import PaginationMeta


class AuditService:
    def __init__(self, repository: AuditLogRepository):
        self.repository = repository

    # -------------------------
    # RECORD AN ACTION
    # -------------------------
    def record(
        self,
        *,
        actor: User | None,
        action: str,
        resource_type: str,
        resource_id: str | None = None,
        summary: str,
        before: dict | None = None,
        after: dict | None = None,
        ip_address: str | None = None,
    ) -> AuditLog:
        log = AuditLog(
            actor_id=actor.id if actor else None,
            actor_email=actor.email if actor else None,
            actor_name=actor.name if actor else None,
            action=action,
            resource_type=resource_type,
            resource_id=str(resource_id) if resource_id else None,
            summary=summary,
            before_data=before,
            after_data=after,
            ip_address=ip_address,
        )
        return self.repository.create(log)

    # -------------------------
    # GET LOGS (ADMIN)
    # -------------------------
    def get_logs(
        self,
        resource_type: str | None = None,
        action: str | None = None,
        actor_email: str | None = None,
        search: str | None = None,
        page: int = 1,
        page_size: int = 20,
    ) -> dict:
        logs, total, total_pages = self.repository.get_logs(
            resource_type=resource_type,
            action=action,
            actor_email=actor_email,
            search=search,
            page=page,
            page_size=page_size,
        )

        return {
            "data": logs,
            "meta": PaginationMeta(
                total=total,
                page=page,
                page_size=page_size,
                total_pages=total_pages,
            ).model_dump(),
        }

File: web/src/services/audit.service.ts
import type { Paginated, QueryParams } from "@/types/api"
import { toPaginated } from "@/types/api"
import { api } from "@/lib/http"

export type AuditAction = "CREATE" | "UPDATE" | "DELETE" | "STATUS_CHANGE"

export interface AuditLogEntry {
  id: string
  actorId: string | null
  actorEmail: string | null
  actorName: string | null
  action: AuditAction
  resourceType: string
  resourceId: string | null
  summary: string
  beforeData: Record<string, unknown> | null
  afterData: Record<string, unknown> | null
  ipAddress: string | null
  createdAt: string
}

interface RawAuditLog {
  id: string
  actor_id: string | null
  actor_email: string | null
  actor_name: string | null
  action: AuditAction
  resource_type: string
  resource_id: string | null
  summary: string
  before_data: Record<string, unknown> | null
  after_data: Record<string, unknown> | null
  ip_address: string | null
  created_at: string
}

export interface AuditLogsQuery extends QueryParams {
  resourceType?: string
  action?: AuditAction
  search?: string
}

function mapAuditLog(raw: RawAuditLog): AuditLogEntry {
  return {
    id: raw.id,
    actorId: raw.actor_id,
    actorEmail: raw.actor_email,
    actorName: raw.actor_name,
    action: raw.action,
    resourceType: raw.resource_type,
    resourceId: raw.resource_id,
    summary: raw.summary,
    beforeData: raw.before_data,
    afterData: raw.after_data,
    ipAddress: raw.ip_address,
    createdAt: raw.created_at,
  }
}

export const auditService = {
  async getLogs(params?: AuditLogsQuery): Promise<Paginated<AuditLogEntry>> {
    const { data, meta } = await api.get<RawAuditLog[]>("/audit-logs", {
      resource_type: params?.resourceType,
      action: params?.action,
      search: params?.search,
      page: params?.page,
      page_size: params?.pageSize,
    })
    return toPaginated(data.map(mapAuditLog), meta)
  },
}
